### ISSUE TITLE
Create a static analysis to detect mutable borrow GC hazards

### ISSUE THREAD

**Issue by jdm** [servo/servo#33140]
https://github.com/servo/servo/pull/33133 contains examples of borrow hazards discovered when a GC occurs while a DOM object's field has a mutable borrow outstanding. It _should_ be possible to detect these automatically at compilation time if we can accurately answer the question "can this function call perform a GC?"

We could build up to this iteratively:
1. create a `struct CanGc(());` which can be copied freely
2. any code in Servo that calls a method from mozjs/rust-mozjs must have a `CanGc` argument passed into it, indicating that it calls a method which can perform a GC operation
3. the compiler now forces all code that calls these methods to pass a `CanGc` argument, so each caller must also receive a new `CanGc` argument
4. step 3 repeats all the way up the stack to code that is automatically called by generated bindings or the main script event loop

Once this data is available, it _should_ be possible to write an analysis pass as a new linter (similar to crown) which uses a dataflow analysis to find any function calls containing a `CanGc` argument that occur within the lifetime of a `borrow_mut()` return value. Even without this static analysis, though, the presence of the arguments will make it easier to visually determine if there are hazards present.

**Comment by sagudev** [comment]
> should be possible to write an analysis pass as a new linter (similar to crown)

Actually I think this should live in crown, but as separate lints, as I do not think we can use multiple rustc drivers at once.

> any code in Servo that calls a method from mozjs/rust-mozjs must have a CanGc argument passed into it

We could enforce this by exposing methods in mozjs using CanGc argument (although I am not sure all mozjs functions can trigger GC).

**Comment by jdm** [comment]
Agreed! It could incrementally work down the stack into rust-mozjs as well for better accuracy.

**Comment by jdm** [comment]
https://github.com/servo/servo/issues/33926 is an example of one hazard that was found by code inspection based on the new CanGc annotations, and then verified by running with GC zeal turned up to maximum in a debug mozjs build.

**Comment by sagudev** [comment]
I was wondering are we sure that can_gc argument even if not used gets optimized out, if not might not be zero cost safety check. I think making `#[crown::cangc]` attribute and adding it to the function would be more correct/easier from crown's standpoint, but it will not be helpful help until we get lint, while current can_gc arguments are already useful.

NOTE: Crown lint will be similar to https://rust-lang.github.io/rust-clippy/master/index.html#/borrow_mut.

I can write the lint if you are not already working on it.

SIDE-NOTE: How do we know which SM methods can cause GC? I think it's all of them (that takes cx), except if otherwise noted.

Also we need to think how this will all connect to existing nogc machinery: https://searchfox.org/mozilla-central/source/js/public/GCAPI.h#1036

**Comment by jdm** [comment]
I am not already working on the lint.

**Comment by jdm** [comment]
> SIDE-NOTE: How do we know which SM methods can cause GC? I think it's all of them (that takes cx), except if otherwise noted.

It's less clear than that, sadly. For example, `ReadableStreamUpdateDataAvailableFromSource` can trigger a GC, while `ReadableStreamIsReadable` can't, but they both take a cx argument.

**Comment by sagudev** [comment]
Creating this lint will be harder than I though, refmut across await lint is very different than what we need (I think we will need visitor).

**Comment by gterzian** [comment]
Where do you track functions that need to be marked as `can_gc`? 

I think `enter_realm` should.

**Comment by jdm** [comment]
It's all manual at this point.

**Comment by gterzian** [comment]
What I mean is: do we have an issue where we keep track of functions that still need to have a `can_gc` argument added to them, so that I don't have to do it myself(seems like very good tasks for new contributors). 

I have opened https://github.com/servo/servo/issues/34573

**Comment by augustebaum** [comment]
Hi all, now that the work on the `reflect_dom_object` callers is done, what would be good to do next? Is there still "mechanical" work that can be done by new contributors, or is expert knowledge needed to know where adding the `can_gc` argument is appropriate?

**Comment by jdm** [comment]
There's still lots of mechanical work from #34573 and existing users of CanGc::note (https://github.com/search?q=repo%3Aservo%2Fservo%20CanGc%3A%3Anote&type=code). I am not a good person to advise on actually creating the static analysis pass right now; I'm only good at spotting current violations of the rules.

**Comment by jdm** [comment]
The rule of thumb is to push CanGc into arguments recursively until you encounter:
* A trait boundary (ie. the code calling CanGc::note is a trait method)
* An async task (the compiler will stop you here)

**Comment by augustebaum** [comment]
Thanks, that's exactly what I was looking for! Working on a PR now.

**Comment by augustebaum** [comment]
What about unsafe extern functions like https://github.com/servo/servo/blob/main/components/script/dom/promise.rs#L332? For now I'm assuming this is similar to the trait boundary case?

**Comment by jdm** [comment]
Ah, yes, extern functions are another boundary case.

**Comment by sagudev** [comment]
GC always happens in same thread, right? Why don't we just use unsafe and allow borrow in tracer (even in presence of mutable borrow). Basically I am wondering why we do not use something akin to https://doc.servo.org/src/script/dom/bindings/cell.rs.html#64-67 in trace impl of refcell.

**Comment by sagudev** [comment]
> GC always happens in same thread, right? Why don't we just use unsafe and allow borrow in tracer (even in presence of mutable borrow). Basically I am wondering why we do not use something akin to https://doc.servo.org/src/script/dom/bindings/cell.rs.html#64-67 in trace impl of refcell.

That would not work as GC is also moving objects.

**Comment by sagudev** [comment]
I think we could use https://github.com/servo/mozjs/issues/520#issuecomment-2702746832 to prevent borrow GC hazards in compile time (more specifically borrow_mut would borrow cx, thus preventing any GC happening while borrow_mut lives). For this we would need to be passing cx everywhere and make sure we handle it safely, but I think that's doable (in codegen) and it would allow to lean into cx for more safety nets (mostly GC related).

**Comment by gterzian** [comment]
I think the `CanGc` for [`ScriptThread::handle_msgs`](https://github.com/servo/servo/blob/c6f61e6b6e9fca33da7655b1a8aa1ebce584f40d/components/script/script_thread.rs#L1341) is too coarse grained, because that method handles many messages that do not Gc, and that instead a finer grained `CanGC::note` should be used for message handling that can gc, which are the ones that currently are passed a `can_gc` argument.

**Comment by jdm** [comment]
> I think the `CanGc` for [`ScriptThread::handle_msgs`](https://github.com/servo/servo/blob/c6f61e6b6e9fca33da7655b1a8aa1ebce584f40d/components/script/script_thread.rs#L1341) is too coarse grained, because that method handles many messages that do not Gc, and that instead a finer grained `CanGC::note` should be used for message handling that can gc, which are the ones that currently are passed a `can_gc` argument.

I'm sympathetic to this view, but this is exactly what CanGc means. If the caller of ScriptThread::handle_msgs has an outstanding mutable borrow, it's important to know that calling handle_msgs can trigger a GC, even if it's not the most common result.

**Comment by sagudev** [comment]
> > GC always happens in same thread, right? Why don't we just use unsafe and allow borrow in tracer (even in presence of mutable borrow). Basically I am wondering why we do not use something akin to https://doc.servo.org/src/script/dom/bindings/cell.rs.html#64-67 in trace impl of refcell.
> 
> That would not work as GC is also moving objects.

Or maybe it would be OK as servo dom object are non-movable as they are allocated via Box. I am mostly wondering in the context of https://github.com/servo/servo/issues/42001#issuecomment-3768013796. There might be some aliasing problems tho. This is definitely something worth exploring/researching as it would allow as to avoid borrowing hazards all together.

**Comment by sagudev** [comment]
> There might be some aliasing problems tho.

It's okay: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=a5430139c250077dc0b894a79727b9b0

the only problem would be if we would mutate anything when tracing, but that does not happen.

**Comment by sagudev** [comment]
ah, bad example, it's okay to hold Ref or RefMut across trace/GC but not `&mut` https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d05c200e8ef5b691a885b28131ce69ef

so we can still use this fact to somehow relax the rules and avoid doing many borrow/borrow_mut (that is we add &JSContext into RefMut deref_mut)

**Comment by sagudev** [comment]
Closing this as the effort is replaced by passing down `&mut JSContext`, which has same role but is even more powerful and will allow to detect gc borrow hazards without any additions to crown (using only rust type system).
